### PR TITLE
frank/dlob

### PR DESCRIPTION
**Pull request by soundsonacid**

### CODE REVIEW

**Review by jordy25519 — COMMENTED** [review]
1. Assuming we will rebuild the node list frequently the Box allocations (and deallocations) will become expensive

Suggest using something like https://crates.io/crates/typed-arena or an object pool to reuse allocations

2. not sure the trait and distinct structs for each node type is helpful?
```rust
struct DlobNode {
    order: Order,
    user_account: Account,
    kind: NodeKind,
    next: Option<*mut DlobNode>
    prev: Option<*mut DlobNode>
}
```

3. can't comment on the use of the raw pointers as I haven't used them much

File: src/dlob/market.rs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+            NodeType::TakingLimit => &self.taking_limit_orders,
+            NodeType::Market => &self.market_orders,
+            NodeType::Trigger => &self.trigger_orders,
+            NodeType::VAMM => panic!("VAMM order list not found"),
+        }
+        .clone()
+    }
+}
+
+pub(crate) type Exchange = DashMap<String, DashMap<u16, Market>>;
```
Comment (jordy25519) [inline src/dlob/market.rs:136]: not sure if hashmap makes sense vs. e.g a struct

```rust
struct Exchange {
    perp: DashMap<u16, Market>>,
    spot: DashMap<u16, Market>>,
}
```

File: src/dlob/dlob_node.rs [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+        unimplemented!()
+    }
+}
+
+pub(crate) fn create_node(node_type: NodeType, order: Order, user_account: Pubkey) -> Node {
+    Node::new(node_type, order, user_account)
+}
+
+pub(crate) fn get_order_signature(order_id: u32, user_account: Pubkey) -> String {
+    format!("{}-{}", order_id, user_account.to_string())
```
Comment (jordy25519) [inline src/dlob/dlob_node.rs:301]: not sure about the string key, something like this with `Display`/to_string method may be better

```rust
#[derive(Eq, Hash, PartialEq)]
struct OrderKey {
    user: Pubkey,
    order_id: u32,
}
```

`pub type OrderKey = (Pubkey, u32);` could also work

**Review by jordy25519 — COMMENTED** [review]
lots of work done here! haven't grokked everything yet but left some simple feedback

File: src/dlob/dlob_node.rs (hunk 2 above)
Comment (soundsonacid) [reply src/dlob/dlob_node.rs:301]: i like the string key, i think for simplicity it makes a lot of sense

File: src/dlob/market.rs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        slot: u64,
+    ) -> (Option<&mut Orderlist>, SubType) {
+        let is_inactive_trigger_order = order.must_be_triggered() && !order.triggered();
+
+        let node_type = if is_inactive_trigger_order {
+            NodeType::Trigger
+        } else if is_one_of_variant(
+            &order.order_type,
+            &[
+                OrderType::Market,
```
[the remark this answers is not in the document]
Comment (soundsonacid) [reply src/dlob/market.rs:56]: sure will do down the line, just need this merged in for jit maker rn

**Pull request merged**